packages/overture-schema-validation/README.md

```diff
@@ -0,0 +1,340 @@
+    pattern=r"^[nwr]\d+$",
+    error_message="Invalid OSM ID format: {value}. Must be n123, w123, or r123"
+)
+
+osm_id: Annotated[str, OSMIdConstraint] = Field(..., description="OSM ID")
+```
+
+### Differences from Traditional Validators
+
+#### Replacing @field_validator
+
+Instead of using Pydantic's `@field_validator` decorator, use constraint annotations. This will
+enable support for richer JSON Schema constraints and preservation of constraint logic in generated
+code.
+
+**Before (using @field_validator):**
+
+```python
+class PlaceProperties(BaseModel):
+    country: str = Field(..., description="Country code")
+    language: str = Field(..., description="Language tag")
+    categories: List[str] = Field(..., description="Categories")
+    wikidata_id: Optional[str] = Field(None, description="Wikidata ID")
+
+    @field_validator("country")
+    @classmethod
+    def validate_country_code(cls, v):
+        if not re.match(r"^[A-Z]{2}$", v):
+            raise ValueError("Invalid ISO 3166-1 alpha-2 country code")
+        return v
+
+    @field_validator("language")
+    @classmethod
+    def validate_language_tag(cls, v):
+        if not re.match(r"^[a-z]{2,3}(-[A-Za-z]{2,8})*$", v):
+            raise ValueError("Invalid IETF BCP-47 language tag")
+        return v
+
+    @field_validator("categories")
+    @classmethod
+    def validate_unique_categories(cls, v):
+        if len(v) != len(set(v)):
+            raise ValueError("Categories must be unique")
+        return v
+
+    @field_validator("wikidata_id")
+    @classmethod
+    def validate_wikidata_format(cls, v):
+        if v is not None and not re.match(r"^Q\d+$", v):
+            raise ValueError("Invalid Wikidata identifier format")
+        return v
+```
+
+**After (using constraints):**
+
+```python
+class PlaceProperties(BaseModel):
+    country: Annotated[str, CountryCodeConstraint()] = Field(
+        ..., description="Country code"
+    )
+    language: Annotated[str, LanguageTagConstraint()] = Field(
+        ..., description="Language tag"
+    )
+    categories: Annotated[List[str], UniqueItemsConstraint()] = Field(
+        ..., description="Categories"
+    )
+    # Domain-specific constraints removed for incremental approach
+```
+
+#### Using Constraint-Based Validation
+
+For complex model-level validation, use the mixin-based constraint system:
+
+```python
+from overture.schema.validation.mixin import ConstraintValidatedModel, at_least_one_of
+
+@at_least_one_of("max_speed", "min_speed")
+class SpeedLimitRule(ConstraintValidatedModel, BaseModel):
+    max_speed: Optional[Speed] = None
+    min_speed: Optional[Speed] = None
+```
+
+##### ⚠️ CRITICAL: Inheritance Order Matters
+
+When using `ConstraintValidatedModel`, it **MUST** come first in the inheritance list:
+
+```python
+# ✅ CORRECT - ConstraintValidatedModel first
+class MyModel(ConstraintValidatedModel, BaseModel):
+    pass
+
+# ❌ WRONG - Will not generate JSON Schema metadata
```

Comment:
Any tests we could add to ensure this? Or more of a gotcha when adding new things?

Reply:
Maybe. It's more of a gotcha, but we may be able to throw at runtime if it's in the wrong order (or maybe not; subclasses may not call it as a `super()`, so we'd be out of the loop). I'll investigate more; flagging the gotcha programmatically would be great.